### DISCUSSION TITLE
Howto deploy a stable diffusion pipeline with DJL Serving and only Java

### DISCUSSION THREAD

**Discussion by fafriat** [deepjavalibrary/djl#2698] [Q&A] [answered]
Hi,

I would like to deploy a pipeline like stable diffusion using only Java with DJL Serving.

If I understood well, the only way would be to use a workflow but I don't know if it is possible to configure the iterative part of the pipeline...

Is there any way with Java to implement a pipeline completely like it is possible with python ?

**Comment by zachgk** [comment, marked as answer]
There was a comparable [discussion on our slack channel](https://deepjavalibrary.slack.com/archives/C01AURG857U/p1689018404733969). For future readers, I will copy the discussion here:

## Slack Discussion

Lanking
  [15 days ago](https://deepjavalibrary.slack.com/archives/C01AURG857U/p1689018455191969?thread_ts=1689018404.733969&cid=C01AURG857U)
https://docs.djl.ai/master/examples/docs/stable_diffusion.html

Frank Afriat
  [15 days ago](https://deepjavalibrary.slack.com/archives/C01AURG857U/p1689018751606009?thread_ts=1689018404.733969&cid=C01AURG857U)
Yes I saw it, but my question is related to DJL Serving. That's also the point of my second question to implement completely a model in Java using DJL-Serving (like in Python implementing simply
def handle(inputs: Input)
).


Frank Afriat
  [15 days ago](https://deepjavalibrary.slack.com/archives/C01AURG857U/p1689018873898369?thread_ts=1689018404.733969&cid=C01AURG857U)
By the way, your implementation in Java with DJL is excellent !


Zach Kimberg
  [15 days ago](https://deepjavalibrary.slack.com/archives/C01AURG857U/p1689025114144339?thread_ts=1689018404.733969&cid=C01AURG857U)
The piece you saw was about having recursion defined through the workflow file. But, you are able to use recursion inside a custom WorkflowFunction  and then call the recursive piece from the workflow file


Frank Afriat
  [14 days ago](https://deepjavalibrary.slack.com/archives/C01AURG857U/p1689060269920349?thread_ts=1689018404.733969&cid=C01AURG857U)
Thank you Zach!, that was exactly the answer I was hoping... the problem I am facing is how to retrieve the workflow models from the WorkflowFunction as it doesn't seem available from the WorkflowExecutor ? 
[@Zach Kimberg](https://deepjavalibrary.slack.com/team/UR6L53480)
 (edited) 


Zach Kimberg
  [14 days ago](https://deepjavalibrary.slack.com/archives/C01AURG857U/p1689095875236019?thread_ts=1689018404.733969&cid=C01AURG857U)
We don’t have a way to get a list of all the workflow models right now. What we have typically done is use executor.executeExpression(…) from the WorkflowFunction. With this, you can build up the function to execute in the expression. You just need the model/function name for the first argument. This can either be hard-coded into the WF or passed as an argument into it. Check out [FunctionsApply](https://github.com/deepjavalibrary/djl-serving/blob/master/serving/src/main/java/ai/djl/serving/workflow/function/FunctionsApply.java) for an example

Frank Afriat
  [14 days ago](https://deepjavalibrary.slack.com/archives/C01AURG857U/p1689097433866829?thread_ts=1689018404.733969&cid=C01AURG857U)
Excellent, I will check that ! Thank you very much ! 
[@Zach Kimberg](https://deepjavalibrary.slack.com/team/UR6L53480)

**Comment by fafriat** [comment]
For future readers, I will copy also the discussion I had with Frank Liu:

my question was more precisely: 

**Frank Afriat**
Is it currently possible to implement a pipeline in Java with DJL Serving implementing a simple interface like
`def handle(inputs: Input)`
in Python ? 

**Frank Liu**
the equivalent of `handle(input: Input)` is `Translator` in Java
You can package your `Translator.java` in your model.zip file


**Frank Afriat**
For the Translator, it solves the vae and text encoder part but not the iteration part.


**Frank Liu**
You are right, we don't have clean way to do the iteration, one work around is a create a dummy model (use Identity block), and do everything in the Translator
this is pretty nasty, but should work


**Frank Afriat**
Yes, not so clean but good idea !
